### PR TITLE
support ABS stage

### PR DESCRIPTION
**Pull request by ondrajodas**
Solving - https://keboola.atlassian.net/browse/COM-316

### CODE REVIEW

**Comment by ondrajodas** [comment]
testy prošly, jen se to tu nějak neprojevilo - https://travis-ci.org/github/keboola/db-writer-snowflake/builds/708752756

File: src/Keboola/DbWriter/Snowflake/Test/StageLoader.php [hunk 1]
```diff
@@ -6,8 +6,12 @@
 use Keboola\StorageApi\Client;
 use Keboola\StorageApi\Options\GetFileOptions;
 
-class S3Loader
+class StageLoader
```
Comment (ondrejhlavacek) [inline src/Keboola/DbWriter/Snowflake/Test/StageLoader.php:9]: Tady bych to spíš nazval `StagingStorageLoader`

File: tests/phpunit/SnowflakeTest.php [hunk 2]
```diff
@@ -42,17 +44,17 @@ public function setUp(): void
...
+        $this->stageLoader = new StageLoader($this->dataDir, $this->storageApi);
     }
 
     private function getInputCsv(string $tableId): string
     {
         return sprintf($this->dataDir . '/in/tables/%s.csv', $tableId);
     }
 
-    private function loadDataToS3(string $tableId): array
+    private function loadDataToStage(string $tableId): array
```
Comment (ondrejhlavacek) [inline tests/phpunit/SnowflakeTest.php:55]: `loadDataToStagingStorage`?

File: tests/phpunit/SnowflakeTest.php [hunk 3]
```diff
@@ -42,17 +44,17 @@ public function setUp(): void
             $this->storageApi->dropBucket($bucketId, ['force' => true]);
         }
 
-        $this->s3Loader = new S3Loader($this->dataDir, $this->storageApi);
+        $this->stageLoader = new StageLoader($this->dataDir, $this->storageApi);
```
Comment (ondrejhlavacek) [inline tests/phpunit/SnowflakeTest.php:47]: `stagingStorageLoader`?

File: src/Keboola/DbWriter/Snowflake/Test/StageLoader.php [hunk 4]
```diff
@@ -45,6 +49,21 @@ public function upload(string $tableId): array
             (new GetFileOptions())->setFederationToken(true)
         );
 
+        if (isset($fileInfo['absPath'])) {
+            return [
+                'stage' => self::STAGE_ABS,
```
Comment (ondrejhlavacek) [inline src/Keboola/DbWriter/Snowflake/Test/StageLoader.php:54]: spíš `storage`, `storageType` nebo `stagingStorage`

File: src/Keboola/DbWriter/Snowflake/Test/StageLoader.php [hunk 5]
```diff
@@ -6,8 +6,12 @@
 use Keboola\StorageApi\Client;
 use Keboola\StorageApi\Options\GetFileOptions;
 
-class S3Loader
+class StageLoader
 {
+    public const STAGE_ABS = 'abs';
+
+    public const STAGE_S3 = 's3';
```
Comment (ondrejhlavacek) [inline src/Keboola/DbWriter/Snowflake/Test/StageLoader.php:13]: `STORAGE_S3` a `STORAGE_ABS`?

File: src/Keboola/DbWriter/Snowflake/Application.php [hunk 6]
```diff
@@ -167,6 +168,9 @@ private function getAdapter(array $manifest): IAdapter
         if (isset($manifest['s3'])) {
```
Comment (ondrejhlavacek) [inline src/Keboola/DbWriter/Snowflake/Application.php:168]: použít konstanty z Loaderu?

File: tests/phpunit/SnowflakeTest.php [hunk 7]
```diff
@@ -578,4 +579,16 @@ public function queryTaggingProvider(): array
...
+    private function getAdapter(string $table): IAdapter
+    {
+        $loadFile = $this->loadDataToStage($table);
+        if ($loadFile['stage'] === StageLoader::STAGE_S3) {
+            return new S3Adapter($loadFile['manifest']);
+        } elseif ($loadFile['stage'] === StageLoader::STAGE_ABS) {
+            return new AbsAdapter($loadFile['manifest']);
+        }
+
+        throw new ApplicationException(sprintf('Stage "%s" does not recognized', $loadFile['stage']));
```
Comment (ondrejhlavacek) [inline tests/phpunit/SnowflakeTest.php:592]: `Staging storage type "%s"...`

**Comment by ondrajodas** [comment]
@ondrejhlavacek checkneš to ještě než to mergnu?

File: tests/phpunit/SnowflakeTest.php [hunk 8]
```diff
@@ -578,4 +579,16 @@ public function queryTaggingProvider(): array
...
+    private function getAdapter(string $table): IAdapter
+    {
+        $loadFile = $this->loadDataToStagingStorage($table);
+        if ($loadFile['stagingStorage'] === StagingStorageLoader::STORAGE_S3) {
+            return new S3Adapter($loadFile['manifest']);
+        } elseif ($loadFile['stagingStorage'] === StagingStorageLoader::STORAGE_ABS) {
+            return new AbsAdapter($loadFile['manifest']);
+        }
+
+        throw new ApplicationException(sprintf('Staging storage type "%s" does not recognized', $loadFile['stage']));
```
Comment (ondrejhlavacek) [inline tests/phpunit/SnowflakeTest.php:592]:
```suggestion
        throw new ApplicationException(sprintf('Staging storage type "%s" not recognized', $loadFile['stage']));
```

**Pull request merged**